In [ ]:
!pip install qiskit

In [ ]:
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit.circuit import Parameter, Instruction, Gate
from qiskit.circuit.library import *
import numpy as np
import matplotlib.pyplot as plt

**STANDARD GATES**

*PAULI, HADAMARD & CONTROLLED GATES*

# Standard Gates and Operations in Qiskit

## 1. Reversible and Unitary Quantum Gates

Quantum gates in Qiskit are generally:

- **Unitary**
- **Reversible**

A quantum gate is represented by a unitary matrix $U$.

### Unitary Condition

$$
U^\dagger U = I
$$

Where:

- $U^\dagger$ = conjugate transpose
- $I$ = identity matrix

This ensures:

- probabilities are preserved
- quantum states remain valid

---

## 2. Reversible Operations

Quantum operations can always be undone.

### Example: X Gate

$$
X =
\begin{bmatrix}
0 & 1 \\
1 & 0
\end{bmatrix}
$$

Applying X twice:

$$
X^2 = I
$$

Meaning:

$$
|0\rangle \rightarrow |1\rangle \rightarrow |0\rangle
$$

So the X gate is reversible.

---

# 3. Abstract Operations

Some Qiskit operations describe:

$$
\text{WHAT the operation does}
$$

rather than:

$$
\text{HOW the hardware implements it}
$$

Example:

```python
from qiskit.circuit.library import MCXGate
```

MCX = Multi-Controlled X gate.

---

## 4. MCXGate

MCX flips the target qubit only when all control qubits are $1$.

Example:

```python
MCXGate(3)
```

Means:

- 3 control qubits
- 1 target qubit

Logical action:

$$
|1110\rangle \rightarrow |1111\rangle
$$

---

## 5. Different Decompositions

MCX does not have one fixed hardware implementation.

Qiskit may decompose it differently depending on:

- backend hardware
- optimization level
- available ancilla qubits
- basis gates

Example decomposition:

$$
H \rightarrow CX \rightarrow T \rightarrow T^\dagger \rightarrow CX
$$

Different decompositions can produce the same logical operation.

---

# Structural Operations

Structural operations are predefined circuit architectures.

They already specify an exact circuit structure.

Example:

```python
from qiskit.circuit.library import RealAmplitudes
```

---

## 6. RealAmplitudes

`RealAmplitudes` is not a single gate.

It is a parameterized circuit template.

Example:

```python
ansatz = RealAmplitudes(3)
```

Typical structure:

$$
RY \rightarrow CX \rightarrow RY \rightarrow CX
$$

It usually contains:

- rotation layers
- entanglement layers

---

## 7. Key Difference

### Abstract Operation

Example:

$$
MCXGate
$$

Defines mathematical behavior.

Can have multiple decompositions.

---

### Structural Operation

Example:

$$
RealAmplitudes
$$

Defines a fixed circuit pattern.

Has a unique structure.

---

# Summary Table

| Type | Example | Meaning |
|---|---|---|
| Standard Gate | X, H, CX | Fixed unitary operation |
| Abstract Operation | MCXGate | Logical action with multiple decompositions |
| Structural Operation | RealAmplitudes | Predefined circuit architecture |

---

# Important Concept

## Mathematical Gates

Represent quantum operations directly.

Examples:

- XGate
- HGate
- RXGate
- MCXGate

---

## Structural Circuits

Represent reusable circuit templates.

Examples:

- RealAmplitudes
- EfficientSU2
- QFT

These are higher-level circuit constructions.

In [ ]:
from qiskit.circuit.library import XGate

xgate = XGate()
print(xgate.to_matrix())

[[0.+0.j 1.+0.j]
 [1.+0.j 0.+0.j]]


In [ ]:
# we can add power to the gate

print(xgate.power(1/2).to_matrix())
print(xgate.power(1/3).to_matrix())
print(xgate.power(1/4).to_matrix())

[[0.5+0.5j 0.5-0.5j]
 [0.5-0.5j 0.5+0.5j]]
[[0.75+0.4330127j 0.25-0.4330127j]
 [0.25-0.4330127j 0.75+0.4330127j]]
[[0.85355339+0.35355339j 0.14644661-0.35355339j]
 [0.14644661-0.35355339j 0.85355339+0.35355339j]]


# Quantum Amplitudes are "Spread Out"

In quantum computing, gates do not simply flip values like classical logic.

Instead, they distribute the quantum state into multiple amplitudes.

---

# Key Idea

A qubit state is written as:

$$
|\psi\rangle = a|0\rangle + b|1\rangle
$$

Where:

- $a$ = amplitude of $|0\rangle$
- $b$ = amplitude of $|1\rangle$

These amplitudes are usually complex numbers.

---

# Important Rule

Quantum states preserve:

$$
|a|^2 + |b|^2 = 1
$$

NOT:

$$
a+b=1
$$

This is one of the most important ideas in quantum mechanics.

---

# Example 1 — Hadamard Gate

Applying Hadamard to $|0\rangle$:

$$
|0\rangle
\rightarrow
\frac{1}{\sqrt2}|0\rangle
+
\frac{1}{\sqrt2}|1\rangle
$$

Numerically:

$$
0.707|0\rangle + 0.707|1\rangle
$$

---

## Check the Probabilities

Probability of measuring $|0\rangle$:

$$
|0.707|^2 = 0.5
$$

Probability of measuring $|1\rangle$:

$$
|0.707|^2 = 0.5
$$

Total probability:

$$
0.5 + 0.5 = 1
$$

---

# Example 2 — Square Root of X Gate

The $\sqrt{X}$ gate partially flips the qubit.

Matrix:

$$
\sqrt{X}=
\begin{bmatrix}
0.5+0.5i & 0.5-0.5i \\
0.5-0.5i & 0.5+0.5i
\end{bmatrix}
$$

Applying it once creates a partial rotation.

Applying it twice gives the full X gate:

$$
(\sqrt{X})^2 = X
$$

---

# Example 3 — Adding Complex Amplitudes

Suppose amplitudes are:

$$
0.14644661-0.35355339i
$$

and

$$
0.85355339+0.35355339i
$$

Add real parts:

$$
0.14644661 + 0.85355339 = 1
$$

Add imaginary parts:

$$
-0.35355339i + 0.35355339i = 0
$$

Final result:

$$
1 + 0i = 1
$$

The imaginary parts cancel out.

---

# Intuition

Quantum gates spread the state across amplitudes.

Instead of:

$$
|0\rangle \rightarrow |1\rangle
$$

you often get:

$$
|0\rangle
\rightarrow
a|0\rangle + b|1\rangle
$$

This is called:

- superposition
- amplitude distribution

---

# Core Concept

Quantum computation preserves:

$$
|a|^2 + |b|^2 = 1
$$

which represents total probability.

In [ ]:
# CX gate

print(xgate.control(1))
print(xgate.control(1).to_matrix())

Instruction(name='cx', num_qubits=2, num_clbits=0, params=[])
[[1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 1.+0.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j]]


Get info about the gates


In [ ]:
from qiskit.circuit.library import get_standard_gate_name_mapping

print(get_standard_gate_name_mapping().keys())

# standard or backend basis gate

dict_keys(['id', 'sx', 'x', 'cx', 'rz', 'r', 'c3sx', 'ccx', 'dcx', 'ch', 'cp', 'crx', 'cry', 'crz', 'cswap', 'csx', 'cu', 'cu1', 'cu3', 'cy', 'cz', 'ccz', 'global_phase', 'h', 'p', 'rccx', 'rcccx', 'rx', 'rxx', 'ry', 'ryy', 'rzz', 'rzx', 'xx_minus_yy', 'xx_plus_yy', 'ecr', 's', 'sdg', 'cs', 'csdg', 'swap', 'iswap', 'sxdg', 't', 'tdg', 'u', 'u1', 'u2', 'u3', 'y', 'z', 'delay', 'reset', 'measure'])


# Single-Qubit Standard Gates in Qiskit

---

# HGate

## Hadamard Gate

Creates superposition.

$$
H =
\frac{1}{\sqrt2}
\begin{bmatrix}
1 & 1 \\
1 & -1
\end{bmatrix}
$$

Transforms:

$$
|0\rangle \rightarrow \frac{|0\rangle+|1\rangle}{\sqrt2}
$$

```python
qc.h(0)
```

---

# IGate

## Identity Gate

Does nothing to the qubit.

$$
I =
\begin{bmatrix}
1 & 0 \\
0 & 1
\end{bmatrix}
$$

```python
qc.id(0)
```

---

# PhaseGate

## Rotation Around Z-Axis

Applies a phase rotation.

$$
P(\lambda)=
\begin{bmatrix}
1 & 0 \\
0 & e^{i\lambda}
\end{bmatrix}
$$

```python
qc.p(theta, 0)
```

---

# RGate

## General Rotation Gate

Rotation around axis:

$$
\cos(\phi)X + \sin(\phi)Y
$$

Notation:

$$
R(\theta,\phi)
$$

```python
qc.r(theta, phi, 0)
```

---

# RXGate

## Rotation Around X-Axis

$$
R_X(\theta)=e^{-i\theta X/2}
$$

```python
qc.rx(theta, 0)
```

---

# RYGate

## Rotation Around Y-Axis

$$
R_Y(\theta)=e^{-i\theta Y/2}
$$

```python
qc.ry(theta, 0)
```

---

# RZGate

## Rotation Around Z-Axis

$$
R_Z(\theta)=e^{-i\theta Z/2}
$$

```python
qc.rz(theta, 0)
```

---

# SGate

## S Gate

Equivalent to:

$$
Z^{0.5}
$$

Matrix:

$$
S=
\begin{bmatrix}
1 & 0 \\
0 & i
\end{bmatrix}
$$

```python
qc.s(0)
```

---

# SdgGate

## S-Dagger Gate

Inverse of S gate.

$$
S^\dagger=
\begin{bmatrix}
1 & 0 \\
0 & -i
\end{bmatrix}
$$

```python
qc.sdg(0)
```

---

# SXGate

## Square Root of X Gate

Half X rotation.

$$
(SX)^2 = X
$$

```python
qc.sx(0)
```

---

# SXdgGate

## Inverse Square Root of X Gate

Inverse of SX gate.

```python
qc.sxdg(0)
```

---

# TGate

## T Gate

Equivalent to:

$$
Z^{0.25}
$$

Matrix:

$$
T=
\begin{bmatrix}
1 & 0 \\
0 & e^{i\pi/4}
\end{bmatrix}
$$

```python
qc.t(0)
```

---

# TdgGate

## T-Dagger Gate

Inverse of T gate.

```python
qc.tdg(0)
```

---

# UGate

## General Single-Qubit Rotation

Most general single-qubit gate.

Uses ZYZ Euler angles.

$$
U(\theta,\phi,\lambda)
$$

```python
qc.u(theta, phi, lambda_, 0)
```

---

# U1Gate

## Z-Axis Rotation Gate

Equivalent to phase rotation.

```python
qc.u1(theta, 0)
```

---

# U2Gate

## Rotation Around X+Z Axis

Uses 2 Euler angles.

```python
qc.u2(phi, lambda_, 0)
```

---

# U3Gate

## Generic 3-Parameter Rotation

Most general older single-qubit rotation gate.

```python
qc.u3(theta, phi, lambda_, 0)
```

---

# XGate

## Pauli-X Gate

Quantum NOT gate.

$$
X=
\begin{bmatrix}
0 & 1 \\
1 & 0
\end{bmatrix}
$$

Transforms:

$$
|0\rangle \leftrightarrow |1\rangle
$$

```python
qc.x(0)
```

---

# YGate

## Pauli-Y Gate

Bit flip + phase flip.

$$
Y=
\begin{bmatrix}
0 & -i \\
i & 0
\end{bmatrix}
$$

```python
qc.y(0)
```

---

# ZGate

## Pauli-Z Gate

Phase flip gate.

$$
Z=
\begin{bmatrix}
1 & 0 \\
0 & -1
\end{bmatrix}
$$

Transforms:

$$
|1\rangle \rightarrow -|1\rangle
$$

```python
qc.z(0)
```

---

# Summary Table

| Gate | Purpose |
|---|---|
| H | Superposition |
| I | Identity / no change |
| P | Phase rotation |
| RX | X-axis rotation |
| RY | Y-axis rotation |
| RZ | Z-axis rotation |
| S | Phase gate |
| T | π/4 phase gate |
| SX | Square-root X |
| U | General unitary |
| X | Bit flip |
| Y | Bit + phase flip |
| Z | Phase flip |

In [ ]:
from qiskit.circuit.library import HGate

print(HGate().to_matrix())

[[ 0.70710678+0.j  0.70710678+0.j]
 [ 0.70710678+0.j -0.70710678+0.j]]


# Hadamard Gate Matrix Explanation

---

# Hadamard Gate

The Hadamard gate creates superposition.

Matrix:

$$
H =
\frac{1}{\sqrt2}
\begin{bmatrix}
1 & 1 \\
1 & -1
\end{bmatrix}
$$

Numerically:

$$
\frac{1}{\sqrt2} \approx 0.70710678
$$

So Qiskit prints:

$$
\begin{bmatrix}
0.70710678+0.j & 0.70710678+0.j \\
0.70710678+0.j & -0.70710678+0.j
\end{bmatrix}
$$

---

# What Does `+0.j` Mean?

Qiskit stores quantum matrices as complex numbers.

Example:

```python
0.70710678 + 0.j
```

means:

$$
0.70710678 + 0i
$$

So the imaginary part is simply zero.

---

# Purpose of Hadamard Gate

The Hadamard gate transforms basis states into superposition states.

---

## Transformation of $|0\rangle$

$$
|0\rangle
\rightarrow
\frac{|0\rangle + |1\rangle}{\sqrt2}
$$

Vector form:

$$
\frac1{\sqrt2}
\begin{bmatrix}
1 \\
1
\end{bmatrix}
$$

---

## Transformation of $|1\rangle$

$$
|1\rangle
\rightarrow
\frac{|0\rangle - |1\rangle}{\sqrt2}
$$

Vector form:

$$
\frac1{\sqrt2}
\begin{bmatrix}
1 \\
-1
\end{bmatrix}
$$

---

# Basis States

$$
|0\rangle =
\begin{bmatrix}
1 \\
0
\end{bmatrix}
$$

$$
|1\rangle =
\begin{bmatrix}
0 \\
1
\end{bmatrix}
$$

---

# Important Rule

The columns of a quantum gate matrix represent:

$$
\text{Transformed basis states}
$$

---

# First Column

The first column is:

$$
H|0\rangle
=
\frac1{\sqrt2}
\begin{bmatrix}
1 \\
1
\end{bmatrix}
$$

which becomes:

$$
\begin{bmatrix}
0.707 \\
0.707
\end{bmatrix}
$$

---

# Second Column

The second column is:

$$
H|1\rangle
=
\frac1{\sqrt2}
\begin{bmatrix}
1 \\
-1
\end{bmatrix}
$$

which becomes:

$$
\begin{bmatrix}
0.707 \\
-0.707
\end{bmatrix}
$$

---

# Final Hadamard Matrix

Combining both columns:

$$
H=
\begin{bmatrix}
0.707 & 0.707 \\
0.707 & -0.707
\end{bmatrix}
$$

---

# Why Divide by $\sqrt2$?

Without normalization:

$$
\begin{bmatrix}
1 \\
1
\end{bmatrix}
$$

has magnitude:

$$
\sqrt{1^2 + 1^2}
=
\sqrt2
$$

Quantum states must have total probability equal to 1.

So divide by $\sqrt2$:

$$
\left(\frac1{\sqrt2}\right)^2
+
\left(\frac1{\sqrt2}\right)^2
=
0.5 + 0.5
=
1
$$

---

# Core Concept

The Hadamard matrix is designed to:

- create equal superposition
- preserve normalization
- keep states orthogonal
- remain unitary


Quantum computing uses **complex numbers** to represent:

- phase
- interference
- rotations
- wave behavior

---

# Complex Number Form

Python uses:

$$
a + bj
$$

where:

$$
j = \sqrt{-1}
$$

Example:

```python
0.707 + 0.j
```

means:

$$
0.707 + 0i
$$

---

# Why Qiskit Shows `+0.j`

Qiskit stores quantum matrices using complex datatype.

Even real numbers are stored as:

```python
0.707 + 0.j
```

for consistency with quantum operations.

---

# Important Concept

Quantum states use:

$$
\textbf{complex amplitudes}
$$

NOT ordinary probabilities.

Probabilities are calculated using:

$$
P = |a|^2
$$

---

# Phase Information

These states are different:

$$
|0\rangle + |1\rangle
$$

$$
|0\rangle - |1\rangle
$$

The minus sign changes the phase.

Phase affects interference and computation.

---

# Euler Formula

Complex numbers describe quantum rotations:

$$
e^{i\theta} = \cos(\theta) + i\sin(\theta)
$$

Used in:

- rotation gates
- phase gates
- Bloch sphere rotations

---

# Core Idea

Complex numbers allow quantum systems to:

- store phase
- create interference
- behave like waves
- perform quantum rotations
- remain unitary

#[Phase (video)](https://youtu.be/oU3terLeU7A?si=4xqoWBcQXXkroB9N)

HGate : Single-qubit Hadamard gate.

IGate : Identity gate.

PhaseGate : Single-qubit rotation about the Z axis.

RGate : Rotation by angle θ around the cos⁡(ϕ)x+sin⁡(ϕ)y axis R(θ,ϕ).

RXGate : Single-qubit rotation about the X axis.

RYGate : Single-qubit rotation about the Y axis.

RZGate : Single-qubit rotation about the Z axis.

SGate : Single qubit S gate (Z**0.5).

SdgGate : Single qubit S-adjoint gate (~Z**0.5).

SXGate : The single-qubit Sqrt(X) gate (SX).

SXdgGate : The inverse single-qubit Sqrt(X) gate.

TGate : Single qubit T gate (Z**0.25).

TdgGate : Single qubit T-adjoint gate (~Z**0.25).

UGate : Generic single-qubit rotation in terms of ZYZ Euler angles.

U1Gate : Single-qubit rotation about the Z axis.

U2Gate : Single-qubit rotation about the X+Z axis.

U3Gate : Generic single-qubit rotation gate with 3 Euler angles.

XGate : The single-qubit Pauli-X gate (σx​).

YGate : The single-qubit Pauli-Y gate (σy​).

ZGate : The single-qubit Pauli-Z gate (σz​).